src/librustc/util/ppaux.rs

```diff
@@ -907,6 +907,27 @@ impl<'tcx> fmt::Display for ty::TypeVariants<'tcx> {
             }
             TyTrait(ref data) => write!(f, "{}", data),
             ty::TyProjection(ref data) => write!(f, "{}", data),
+            ty::TyAnon(def_id, substs) => {
+                ty::tls::with(|tcx| {
+                    // Grab the "TraitA + TraitB" from `impl TraitA + TraitB`,
```

Comment:
> The precedence is this way for convenience - unlike trait objects, there's a prefix (impl), which solves all ambiguity save for what you just mentioned.

I don't think there was an more-or-less "ambiguity" before, it's just that we had a weird set of precedence rules. I think the reason I find the precedence of `impl` relative to `+` surprising that, in general, we typically give keyword operators the same precedence as other unary operators and it usually higher than binary operators. Hence `box 3 + 2` parses as `(box 3) + 2`. (Admittedly, though, I have [complained about this in the past](https://github.com/rust-lang/rust/issues/21192).)

In any case, I continue to find it inconsistent that the `Send` is associated with the trait in `impl Trait + Send` but with the type parameter `F` in `F: Fn() Foo -> Trait + Send`. (Even if I agree with your argument that one would not necessarily combine them... yet...)

Seems like maybe we should raise it on the tracking issue?